File: examples/java-checker-framework/src/main/java/org/example/arraymap/ArrayMap.java
package org.example.arraymap;

import org.checkerframework.checker.index.qual.IndexOrHigh;
import org.checkerframework.checker.index.qual.NonNegative;
import org.checkerframework.checker.index.qual.SameLen;
import org.checkerframework.dataflow.qual.Pure;

public class ArrayMap<K, V> {       //> ArrayMap::constructor p=(3,2,3/3) r=none loc=4
    private final K[] keys;
    private final V @SameLen("this.keys")[] values;
    private @IndexOrHigh("this.keys") int currSize;

    @SuppressWarnings("unchecked")
    public ArrayMap(@NonNegative int capacity) {    //> ArrayMap::aux-constructor p=(1,1,1/1) r=none loc=5
        this.keys = (K[]) new Object[capacity];
        this.values = (V[]) new Object[this.keys.length];
        this.currSize = 0;
    }

    public @Pure @NonNegative int capacity(){       //> ArrayMap::capacity p=(0,0,0/0) r=(1,1/1) loc=3  --  we don't count the pure annotation
        return this.keys.length;
    }

    public @IndexOrHigh("this.keys") int currentSize() {    //> ArrayMap::currentSize p=(0,0,0/0) r=(1,2/2) loc=3
        return this.currSize;
    }

    public V get(K k) {                             //> ArrayMap::get p=(1,0,0/0) r=(0,0/0) loc=8
        var idx = ArrayUtils.indexOf(keys, k);
        if (idx == -1) {
            return null;
        } else {
            return this.values[idx];
        }
    }

    public boolean put(K k, V v) {                  //> ArrayMap::put p=(2,0,0/0) r=(0,0/0) BUG REPORTED loc=18
        if (k == null || v == null) {
            throw new IllegalArgumentException();
        }
        var preSize = currentSize();
        if (preSize < this.keys.length) {
            // ERROR: should be < instead of <=
            for (int i = 0; i <= this.keys.length; i++) {
                if (this.keys[i] == null) {
                    this.keys[i] = k;
                    this.values[i] = v;
                    break;
                }
            }
            this.currSize = preSize + 1;
            return true;
        }
        return false;
    }

    public boolean remove(K k) {                //> ArrayMap::remove p=(1,0,0/0) r=(0,0/0) loc=14
        var preSize = currentSize();
        if (preSize > 0) {
            var idx = ArrayUtils.indexOf(this.keys, k);
            if (idx == -1) {
                return false;
            }
            this.keys[idx] = null;
            this.values[idx] = null;
            this.currSize = preSize - 1;
            return true;
        }
        return false;
    }

}

File: examples/scala/src/main/scala/arraymap/ArrayMap.scala
package arraymap

import scala.reflect.ClassTag
import scala.util.boundary

class ArrayMap[K: ClassTag, V: ClassTag] private( //> ArrayMap::constructor loc=5
                                                  private val keys: Array[Option[K]],
                                                  private val values: Array[Option[V]],
                                                  private var _currSize: Int
                                                ) {

  def this(capacity: Int) = this(new Array(capacity), new Array(capacity), 0) //> ArrayMap::aux-constructor loc=1

  export keys.length as capacity //> ArrayMap::capacity loc=1

  def currentSize: Int = _currSize //> ArrayMap::currentSize loc=1

  def get(k: K): Option[V] = { //> ArrayMap::get loc=6
    val idx = keys.indexOf(Some(k))
    Option.when(idx != -1) {
      values(idx)
    }.flatten
  }

  def put(k: K, v: V): Boolean = { //> ArrayMap::put loc=17
    val preSize = currentSize
    val canAdd = preSize < capacity
    if (canAdd) {
      boundary {
        // ERROR: should be until instead of to (or better: for i <- keys.indices)
        for (i <- 0 to keys.length) {
          if (keys(i) == null) {
            keys(i) = Some(k)
            values(i) = Some(v)
          }
        }
      }
      _currSize = preSize + 1
    }
    canAdd
  }

  def remove(k: K): Boolean = { //> ArrayMap::remove loc=13
    val preSize = currentSize
    if (preSize > 0) {
      val idx = keys.indexOf(Some(k))
      if idx == -1 then false
      else {
        keys(idx) = None
        values(idx) = None
        _currSize = preSize - 1
        true
      }
    } else false
  }

}

extension [T](a: Array[T]) def indexOf(elem: T): Int = {    //> Array::indexOf loc=10
  var i = 0
  while (i < a.length) {
    if (a(i) == elem) {
      return i
    }
    i += 1
  }
  -1
}
